specs/commandline/plugin.md

```diff
@@ -41,51 +42,160 @@ Aliases:
 ### notation plugin install
 
 ```text
-Installs a plugin
+Install a plugin
 
 Usage:
-  notation plugin install [flags] <plugin path>
+  notation plugin install [flags] <plugin_source>
 
 Flags:
   -h, --help                    help for install
   -f, --force                   force the installation of a plugin
+      --checksum string         must match SHA256 of the plugin source               
 
 Aliases:
   install, add
 ```
 
-### notation plugin remove
+### notation plugin upgrade
 
 ```text
-Removes a plugin
+Upgrade a plugin
 
 Usage:
-  notation plugin remove [flags] <plugin name>
+  notation plugin upgrade [flags] <plugin_source>
 
 Flags:
-  -h, --help          help for remove
+  -h, --help                    help for upgrade
+      --checksum string         must match SHA256 of the plugin source                 
+```
+
+### notation plugin uninstall
+
+```text
+Uninstall a plugin
 
+Usage:
+  notation plugin uninstall [flags] <plugin_name>
+
+Flags:
+  -h, --help          help for remove
+  -y, --yes           do not prompt for confirmation
 Aliases:
-  remove, rm, uninstall, delete
+  uninstall, remove, rm
 ```
 
 ## Usage
 
-### Install a plugin
+## Install a plugin 
+
+### Install a plugin from file system
+
+Install a Notation plugin from file system. The checksum validation is optional for this case.
 
 ```shell
-notation plugin install <plugin path>
+$ notation plugin install <file_path>
```

Comment:
Can we please specify the file supported file formats and install behavior for them ?
For Example:

1. `.zip`: The ZIP file should house the plugin in the root directory. If the ZIP archive contains multiple files, Notation will locate an executable file associated with the plugin and copy it, ignoring any other files?
2. `.tar.gz`: Similar for .tar.gz
3. `.tar`: Should we even support?
4. `.gz`: Should we even support?
5. All other format, Notation will treat them as executable?

^ Notation will use file extension to determine how to unpack file/archive ?

Reply:
@priteshbandi These questions are related to implementation though. Since you asked, here are the current design:
1. For `.zip`, Notation goes through the zip file -> find an executable file -> validate its name (must be `notatin-{plugin-name}`) -> check if plugin already exists (if --force is not set) -> extract the executable file into a `tmp` dir (regardless the final result is a success or a failure, this `tmp` dir will be removed in the end) -> validate the executable file is indeed a valid Notation plugin by calling its `get-plugin-metada` -> at this point, copy the file into correct Notation dir with all parent dir created along the way. A special note for `.zip` file, a normal file such as `LICENSE` can be treated as executable too. That's why we cannot directly use whatever executable file Notation finds. 
2. For `.tar.gz`, Notation will first detect that a `.gz` file is passed in. Then, Notation will require this `.gz` file compressed a `.tar` file in it. If yes, then apply similar logic as `.zip`.
3. `.tar` alone is not supported.
4. `.gz` alone is not supported.
5. All other format is not very different from `.tar` or `.gz`, they are not supported.

And yes, Notation has the capability of telling apart `.zip` and `.tar.gz`, but not using file extension. Instead, it considers at most the first 512 bytes of data and determine the file type.